pylearn2/train_extensions/wmape_channel.py

```diff
@@ -0,0 +1,116 @@
+"""
+TrainExtension subclass for calculating Weighted Mean Average Percentage
+Error scores for regression models on monitoring dataset(s), reported via
+monitor channels.
+"""
+
+__author__ = "Junbo chen"
+__copyright__ = "Copyright 2015, Alibaba Group"
+__license__ = "3-clause BSD"
+__maintainer__ = "Junbo Chen"
+
+import numpy as np
+
+import theano
+from theano import gof, config
+from theano import tensor as T
+
+from pylearn2.train_extensions import TrainExtension
+
+
+class WMapeOp(gof.Op):
+    """
+    Theano Op calculate wmape score.
+
+    Parameters
+    ----------
+    name : str, optional (default 'wmape')
+        Name of this Op.
+    use_c_code : WRITEME
+    """
+    def __init__(self, name='wmape', use_c_code=theano.config.cxx):
+        super(WMapeOp, self).__init__(use_c_code)
+        self.name = name
+
+    def make_node(self, y_true, y_score):
+        """
+        Calculate WMAPE score.
+
+        Parameters
+        ----------
+        y_true : tensor_like
+            Target regression values.
+        y_score : tensor_like
+            Predicted regression values.
+        """
+        y_true = T.as_tensor_variable(y_true)
+        y_score = T.as_tensor_variable(y_score)
+        output = [T.scalar(name=self.name, dtype=config.floatX)]
+        return gof.Apply(self, [y_true, y_score], output)
+
+    def perform(self, node, inputs, output_storage):
+        """
+        Calculate WMAPE score.
+
+        Parameters
+        ----------
+        node : Apply instance
+            Symbolic inputs and outputs.
+        inputs : list
+            Sequence of inputs.
+        output_storage : list
+            List of mutable 1-element lists.
+        """
+        y_true, y_score = inputs
+        wmape = np.sum(abs(y_true - y_score)) / np.sum(abs(y_true))
+        output_storage[0][0] = theano._asarray(wmape, dtype=config.floatX)
```

Comment:
This would also make sure the computation is done on the GPU if you are using one, rather than transferring it over to CPU to be able to use numpy.